finagle-redis/src/main/scala/com/twitter/finagle/redis/SubscribeClient.scala

```diff
@@ -0,0 +1,108 @@
+package com.twitter.finagle.redis
+
+import com.twitter.finagle.Service
+import com.twitter.finagle.redis.protocol._
+import com.twitter.finagle.redis.util.StringToChannelBuffer
+import org.jboss.netty.buffer.ChannelBuffer
+import com.twitter.util.Future
+import com.twitter.finagle.ServiceFactory
+import com.twitter.finagle.redis.util.CBToString
+import com.twitter.util.Futures
+import com.twitter.finagle.util.DefaultTimer
+import com.twitter.conversions.time._
+import com.twitter.util.Time
+import com.twitter.util.TimerTask
+import com.twitter.finagle.ServiceClosedException
+
+object SubscribeClient {
+
+  type MessageHandler = (ChannelBuffer, ChannelBuffer) => Unit
+  type PMessageHandler = (ChannelBuffer, ChannelBuffer, ChannelBuffer) => Unit
+
+  /**
+   * Construct a sentinel client from a single host.
+   * @param host a String of host:port combination.
+   */
+  def apply(host: String): SubscribeClient = {
+    new SubscribeClient(com.twitter.finagle.Redis.Subscribe.newService(host))
+  }
+
+  object MessageBytes {
+    val SUBSCRIBE = StringToChannelBuffer("subscribe")
+    val UNSUBSCRIBE = StringToChannelBuffer("unsubscribe")
+    val PSUBSCRIBE = StringToChannelBuffer("psubscribe")
+    val PUNSUBSCRIBE = StringToChannelBuffer("punsubscribe")
+    val MESSAGE = StringToChannelBuffer("message")
+    val PMESSAGE = StringToChannelBuffer("pmessage")
+  }
+}
+
+class SubscribeClient(service: Service[SubscribeCommand, Unit]) extends SubscribeListener {
+
+  import SubscribeClient._
+
+  def subscribe(channels: Seq[ChannelBuffer])(handler: MessageHandler): Future[Unit] =
+    doRequest(Subscribe(channels, this))
+      .onSuccess(_ => channels.foreach(msgHandlers.put(_, handler)))
+
+  def pSubscribe(patterns: Seq[ChannelBuffer])(handler: PMessageHandler): Future[Unit] =
+    doRequest(PSubscribe(patterns, this))
+      .onSuccess(_ => patterns.foreach(pMsgHandlers.put(_, handler)))
+
+  def unsubscribe(channels: Seq[ChannelBuffer]): Future[Unit] =
+    doRequest(Unsubscribe(channels, this))
+      .onSuccess(_ => channels.foreach(msgHandlers.remove(_)))
+
+  def pUnsubscribe(patterns: Seq[ChannelBuffer]): Future[Unit] =
+    doRequest(PUnsubscribe(patterns, this))
+      .onSuccess(_ => patterns.foreach(pMsgHandlers.remove(_)))
+
+  /**
+   * Releases underlying service object
+   */
+  def release() = {
+    service.close()
+  }
+
+  def doRequest(cmd: SubscribeCommand) = {
+    service(cmd)
+  }
+
+  import scala.collection.mutable.HashMap
+  import com.twitter.finagle.redis.protocol._
+  import com.twitter.finagle.redis.protocol.{ Subscribe => SubscribeCmd }
+
+  private val msgHandlers = HashMap[ChannelBuffer, MessageHandler]()
+  private val pMsgHandlers = HashMap[ChannelBuffer, PMessageHandler]()
+
+  def onException(e: Throwable) {
```

Comment:
be explicit here about the type you're returning: `: Unit = {`.  Are you sure you don't want to return `: Future[Unit]`?